jsign-crypto/src/main/java/net/jsign/PrivateKeyUtils.java

```diff
@@ -1,154 +1,161 @@
-/**
- * Copyright 2017 Emmanuel Bourg
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
-package net.jsign;
-
-import java.io.File;
-import java.io.FileReader;
-import java.io.IOException;
-import java.lang.reflect.Field;
-import java.security.KeyException;
-import java.security.PrivateKey;
-import java.util.HashMap;
-import java.util.function.Function;
-
-import org.bouncycastle.asn1.ASN1ObjectIdentifier;
-import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
-import org.bouncycastle.jce.provider.BouncyCastleProvider;
-import org.bouncycastle.openssl.PEMDecryptorProvider;
-import org.bouncycastle.openssl.PEMEncryptedKeyPair;
-import org.bouncycastle.openssl.PEMKeyPair;
-import org.bouncycastle.openssl.PEMParser;
-import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
-import org.bouncycastle.openssl.jcajce.JceOpenSSLPKCS8DecryptorProviderBuilder;
-import org.bouncycastle.openssl.jcajce.JcePEMDecryptorProviderBuilder;
-import org.bouncycastle.operator.InputDecryptorProvider;
-import org.bouncycastle.operator.OperatorCreationException;
-import org.bouncycastle.pkcs.PKCS8EncryptedPrivateKeyInfo;
-import org.bouncycastle.pkcs.PKCSException;
-import sun.misc.Unsafe;
-
-/**
- * Helper class for loading private keys (PVK or PEM, encrypted or not).
- * 
- * @author Emmanuel Bourg
- * @since 2.0
- */
-public class PrivateKeyUtils {
-
-    private PrivateKeyUtils() {
-    }
-
-    /**
-     * Load the private key from the specified file. Supported formats are PVK and PEM,
-     * encrypted or not. The type of the file is inferred from its extension (<code>.pvk</code>
-     * for PVK files, <code>.pem</code> for PEM files).
-     * 
-     * @param file     the file to load the key from
-     * @param password the password protecting the key
-     * @return the private key loaded
-     * @throws KeyException if the key cannot be loaded
-     */
-    public static PrivateKey load(File file, String password) throws KeyException {
-        try {
-            if (file.getName().endsWith(".pvk")) {

```

Comment:
What about keeping the extension based detection and then falling back to the other method?

Reply:
I considered that approach for a while, and realized it wouldn't handle cases where key files have misleading extensions (e.g., a PVK key with a `.pem` extension, or vice versa). Ultimately, I believe that allowing the key to be parsed regardless of the file extension, even if it's misleading, is more user-friendly. That said, I don’t feel strongly about this and am open to changing it if requested :smile: